File: TestC_Code/hnsw.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <limits>
#include <map>
#include <iomanip>
#include <fstream>

using namespace std;

// ---------- Cosine Similarity Function ----------
float cosineSimilarity(const vector<float> &a, const vector<float> &b)
{
    float dot = 0.0, normA = 0.0, normB = 0.0;
    for (size_t i = 0; i < a.size(); ++i)
    {
        dot += a[i] * b[i];
        normA += a[i] * a[i];
        normB += b[i] * b[i];
    }
    return dot / (sqrt(normA) * sqrt(normB));
}

// ---------- Node Structure ----------
struct Node
{
    int id;
    vector<float> embedding; // 🟢 Renamed from `vector`
    map<int, vector<int>> neighbors;
};

// ---------- HNSW Class ----------
class HNSW
{
    vector<Node> nodes;
    int maxLevel;
    int ef;

public:
    HNSW(int maxLevel = 2, int ef = 2) : maxLevel(maxLevel), ef(ef) {}

    void addNode(const vector<float> &vec)
    {
        int id = nodes.size();
        Node node = {id, vec, {}};
        int level = rand() % (maxLevel + 1);

        cout << "\n🔷 Inserting Node " << id << " [";
        for (auto v : vec)
            cout << fixed << setprecision(2) << v << " ";
        cout << "] at level " << level << "\n";

        if (!nodes.empty())
        {
            for (int l = maxLevel; l >= 0; --l)
            {
                int bestId = searchAtLayer(vec, l);
                if (bestId != -1)
                {
                    node.neighbors[l].push_back(bestId);
                    nodes[bestId].neighbors[l].push_back(id);

                    cout << "  ↳ Connected at level " << l
                         << " with Node " << bestId
                         << " (Similarity: "
                         << fixed << setprecision(3)
                         << cosineSimilarity(vec, nodes[bestId].embedding) << ")\n";
                }
            }
        }

        nodes.push_back(node);
    }

    int searchAtLayer(const vector<float> &query, int level)
    {
        float bestSim = -1.0;
        int bestId = -1;
        for (const auto &node : nodes)
        {
            if (node.neighbors.count(level))
            {
                float sim = cosineSimilarity(query, node.embedding);
                if (sim > bestSim)
                {
                    bestSim = sim;
                    bestId = node.id;
                }
            }
        }
        return bestId;
    }

    int search(const vector<float> &query)
    {
        int current = 0;
        float currentSim = cosineSimilarity(query, nodes[current].embedding);

        cout << "\n🔍 Starting search from Node " << current << " [";
        for (auto v : nodes[current].embedding)
            cout << fixed << setprecision(2) << v << " ";
        cout << "] → Similarity: " << fixed << setprecision(3) << currentSim << "\n";

        for (int l = maxLevel; l >= 0; --l)
        {
            cout << "\n⬇️  Level " << l << ": Navigating...\n";
            bool improved = true;
            while (improved)
            {
                improved = false;
                for (int neighborId : nodes[current].neighbors[l])
                {
                    float sim = cosineSimilarity(query, nodes[neighborId].embedding);
                    cout << "    Comparing with Node " << neighborId
                         << " → Similarity: " << fixed << setprecision(3) << sim;

                    if (sim > currentSim)
                    {
                        cout << " ✅ Better! Moving to Node " << neighborId << "\n";
                        current = neighborId;
                        currentSim = sim;
                        improved = true;
                        break;
                    }
                    else
                    {
                        cout << " ❌ Not better.\n";
                    }
                }
            }
        }

        cout << "\n✅ Final Result → Best match: Node " << current
             << " with Similarity: " << fixed << setprecision(4) << currentSim << "\n";

        return current;
    }

    void generateDot(const string &filename)
    {
        ofstream out(filename);
        out << "digraph HNSW {\n";
        out << "  rankdir=LR;\n";
        out << "  node [shape=circle, style=filled, fillcolor=lightyellow];\n";

        for (int l = maxLevel; l >= 0; --l)
        {
            out << "  subgraph cluster_" << l << " {\n";
            out << "    label=\"Level " << l << "\";\n";
            out << "    style=dashed;\n";
            for (const auto &node : nodes)
            {
                if (node.neighbors.count(l))
                {
                    out << "    n" << node.id << " [label=\""
                        << node.id << "\"];\n";
                }
            }
            out << "  }\n";
        }

        for (const auto &node : nodes)
        {
            for (const auto &[level, neighbors] : node.neighbors)
            {
                for (int nId : neighbors)
                {
                    string color = (level == 2 ? "red" : level == 1 ? "blue"
                                                                    : "black");
                    out << "  n" << node.id << " -> n" << nId
                        << " [label=\"L" << level << "\", color=" << color << "];\n";
                }
            }
        }

        out << "}\n";
        out.close();
        cout << "📦 Graphviz DOT file generated: " << filename << endl;
    }
};

// ---------- Main ----------
int main()
{
    srand(time(NULL));
    HNSW hnsw;

    vector<vector<float>> data = {
        {1, 0}, {0.5, 0.5}, {0, 1}, {-1, 0}, {0.6, 0.8}};

    for (const auto &vec : data)
    {
        hnsw.addNode(vec);
    }

    vector<float> query = {0.7f, 0.7f};
    hnsw.search(query);
    hnsw.generateDot("hnsw_graph.dot");

    return 0;
}

File: HNSW/HNSW.js
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

class Node {
  constructor(id, vector) {
    this.id = id;
    this.vector = vector;
    this.neighbors = new Map(); // level -> array of node IDs
  }
}

class HNSW {
  constructor(maxLevel = 2) {
    this.nodes = [];
    this.maxLevel = maxLevel;
  }

  cosineSimilarity(a, b) {
    let dot = 0,
      normA = 0,
      normB = 0;
    for (let i = 0; i < a.length; i++) {
      dot += a[i] * b[i];
      normA += a[i] * a[i];
      normB += b[i] * b[i];
    }
    return dot / (Math.sqrt(normA) * Math.sqrt(normB));
  }

  angleDegrees(similarity) {
    // Clamp to avoid invalid acos
    const clamped = Math.min(1, Math.max(-1, similarity));
    return Math.acos(clamped) * (180 / Math.PI);
  }

  addNode(vector) {
    const id = this.nodes.length;
    const node = new Node(id, vector);
    const level = Math.floor(Math.random() * (this.maxLevel + 1));

    console.log(
      `\n🔷 Inserting Node ${id} [${vector.join(", ")}] at level ${level}`
    );

    for (let l = this.maxLevel; l >= 0; l--) {
      const bestId = this.searchAtLayer(vector, l);
      if (bestId !== -1) {
        if (!node.neighbors.has(l)) node.neighbors.set(l, []);
        node.neighbors.get(l).push(bestId);

        const neighborNode = this.nodes[bestId];
        if (!neighborNode.neighbors.has(l)) neighborNode.neighbors.set(l, []);
        neighborNode.neighbors.get(l).push(id);

        const sim = this.cosineSimilarity(vector, neighborNode.vector);
        const angle = this.angleDegrees(sim);
        console.log(
          `  ↳ Connected at level ${l} with Node ${bestId} (Similarity: ${sim.toFixed(
            3
          )}, Angle: ${angle.toFixed(2)}°)`
        );
      }
    }

    this.nodes.push(node);
  }

  searchAtLayer(query, level) {
    let bestSim = -1;
    let bestId = -1;
    for (const node of this.nodes) {
      if (node.neighbors.has(level)) {
        const sim = this.cosineSimilarity(query, node.vector);
        if (sim > bestSim) {
          bestSim = sim;
          bestId = node.id;
        }
      }
    }
    return bestId;
  }

  async search(query) {
    let current = 0;
    let currentSim = this.cosineSimilarity(query, this.nodes[current].vector);

    console.log(
      `\n🔍 Starting search from Node ${current} [${this.nodes[
        current
      ].vector.join(", ")}] → Similarity: ${currentSim.toFixed(3)}`
    );

    for (let l = this.maxLevel; l >= 0; l--) {
      console.log(`\n⬇️  Level ${l}: Navigating...`);
      let improved = true;
      while (improved) {
        improved = false;
        const neighbors = this.nodes[current].neighbors.get(l) || [];
        for (const neighborId of neighbors) {
          const sim = this.cosineSimilarity(
            query,
            this.nodes[neighborId].vector
          );
          const angle = this.angleDegrees(sim);
          process.stdout.write(
            `    Comparing with Node ${neighborId} → Similarity: ${sim.toFixed(
              3
            )}, Angle: ${angle.toFixed(2)}°`
          );

          await sleep(700); // 💤 delay for visualization
          if (sim > currentSim) {
            console.log(" ✅ Better! Moving.");
            current = neighborId;
            currentSim = sim;
            improved = true;
            break;
          } else {
            console.log(" ❌ Not better.");
          }
        }
      }
    }

    console.log(
      `\n✅ Final Result → Best match: Node ${current} with Similarity: ${currentSim.toFixed(
        4
      )}\n`
    );
    return current;
  }
}

// -------------------------- Demo --------------------------
(async () => {
  const hnsw = new HNSW(2);

  const data = [
    [1, 0],
    [0.5, 0.5],
    [0, 1],
    [-1, 0],
    [0.6, 0.8],
  ];

  for (const vec of data) {
    hnsw.addNode(vec);
  }

  const query = [0.7, 0.7];
  await hnsw.search(query);
})();
